File: src/ClientApplication/PhaserInterop.cs
using Microsoft.JSInterop;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ClientApplication
{
    public interface IPhaserInterop
    {
        void RegisterScene(Scene scene);

        void Start(string containerElement, string title);

        IPhaserSceneInterop InScene(Scene scene);
    }

    public interface IPhaserSceneInterop
    {
        IPhaserSceneInterop AddSprite(string name, string imageName, int x, int y, double scale = 1);

        IPhaserSceneInterop SetSpriteScale(string name, double scale);

        IPhaserSceneInterop SetSpriteCrop(string name, int x, int y, int width, int height);

        IPhaserSceneInterop AddRectangle(int x, int y, int width, int height, string color = "black");

        // TODO Create Sync/Async versions
        IPhaserSceneInterop OnPointerUp(string spriteName, string handlerName);

        void StartScene(string name);

        void StopScene(string name);

        void SwitchToScene(string name);
    }

    public class PhaserSceneInterop : IPhaserSceneInterop
    {
        private readonly IJSInProcessRuntime _jsRuntime;
        private readonly Scene _scene;

        public PhaserSceneInterop(IJSInProcessRuntime jsRuntime, Scene scene)
        {
            _jsRuntime = jsRuntime;
            _scene = scene;
        }

        public IPhaserSceneInterop AddSprite(string name, string imageName, int x, int y, double scale = 1)
        {
            _jsRuntime.InvokeVoid("addSprite", _scene.GetName(), name, imageName, x, y, scale);
            return this;
        }

        public IPhaserSceneInterop SetSpriteScale(string name, double scale)
        {
            _jsRuntime.InvokeVoid("setSpriteScale", _scene.GetName(), name, scale);
            return this;
        }

        public IPhaserSceneInterop SetSpriteCrop(string name, int x, int y, int width, int height)
        {
            _jsRuntime.InvokeVoid("setSpriteCrop", _scene.GetName(), name, x, y, width, height);
            return this;
        }

        public IPhaserSceneInterop AddRectangle(int x, int y, int width, int height, string color)
        {
            _jsRuntime.InvokeVoid("addRectangle", _scene.GetName(), x, y, width, height, color);
            return this;
        }

        public IPhaserSceneInterop OnPointerUp(string spriteName, string handlerName)
        {
            _jsRuntime.InvokeVoid("addHandler", _scene.GetName(), spriteName, handlerName);
            return this;
        }

        public void StartScene(string name)
        {
            _jsRuntime.InvokeVoid("startScene", name);
        }

        public void StopScene(string name)
        {
            _jsRuntime.InvokeVoid("stopScene", name);
        }

        public void SwitchToScene(string name)
        {
            _jsRuntime.InvokeVoid("switchScene", _scene.GetName(), name);
        }
    }

    // public class PhaserRectangleInterop : IPhaserRectangleInterop
    // {
    //     private readonly IJSInProcessRuntime _jsRuntime;
    //     private readonly Scene _scene;
    //     private readonly int _width;
    //     private readonly int _height;

    //     public PhaserRectangleInterop(IJSInProcessRuntime jsRuntime, Scene scene, int width, int height)
    //     {
    //         _jsRuntime = jsRuntime;
    //         _scene = scene;
    //         _width = width;
    //         _height = height;
    //     }

    //     public IPhaserSceneInterop PlaceAt(int x, int y)
    //     {
    //         _jsRuntime.InvokeVoid("addRectangle", _scene.GetName(), _width, _height, x, y);

    //         return new PhaserSceneInterop(_jsRuntime, _scene);
    //     }
    // }

    /// <summary>
    /// Generic client class that interfaces .NET Standard/Blazor with SignalR Javascript client
    /// </summary>
    public class PhaserClient : IPhaserInterop
    {
        private readonly IJSInProcessRuntime _jSRuntime;

        public PhaserClient(IJSRuntime jsRuntime)
        {
            _jSRuntime = (IJSInProcessRuntime)jsRuntime;
        }

        public void RegisterScene(Scene scene)
        {
            _jSRuntime.InvokeVoid("registerScene", scene.GetName(), DotNetObjectReference.Create(scene));
        }

        // public void SwitchScene(Scene from, Scene to)
        // {
        //     _jSRuntime.InvokeVoid("switchScene", from.GetName(), to.GetName());
        // }

        public void Start(string containerElement, string title)
        {
            _jSRuntime.InvokeVoid("startPhaser", containerElement, title);
        }

        public IPhaserSceneInterop InScene(Scene scene)
            => new PhaserSceneInterop(_jSRuntime, scene);

        // public IPhaserClient AddSprite(Scene scene, string sprite, string imageName, int x, int y, double scale, bool interactive)
        // {
        //     _jSRuntime.InvokeVoid("addSprite", scene.GetName(), sprite, imageName, x, y, scale, interactive);
        //     return this;
        // }

        // public IPhaserClient AddHandler(Scene scene, string spriteName, string handlerName)
        // {
        //     _jSRuntime.InvokeVoid("addHandler", scene.GetName(), spriteName, handlerName);
        //     return this;
        // }
    }
}

File: src/ClientApplication/wwwroot/js/phaser.interop.js
const sceneInfos = [];
let currentScene;

// TODO Extract more generic method
function findSprite(sceneName, spriteName) {
    const sceneInfo = sceneInfos[sceneName];
    return sceneInfo.phaserScene.children.list.find(child => {
        return child.type === "Sprite" && child.name === spriteName
    });
}
function findText(sceneName, textName) {
    const sceneInfo = sceneInfos[sceneName];
    return sceneInfo.phaserScene.children.list.find(child => {
        return child.type === "Text" && child.name === textName
    });
}

function addSpriteEventHandler(sceneName, spriteName, eventName, handlerName) {
    const sprite = findSprite(sceneName, spriteName);
    sprite.setInteractive({ pixelPerfect: true });
    sprite.on(eventName, (pointer) => {
        const sceneInfo = sceneInfos[sceneName];
        const eventArgs = { 'spriteName': spriteName, 'spriteX': sprite.x, 'spriteY': sprite.y, 'x': pointer.x, 'y': pointer.y, 'distance': -1 };
        if (pointer.distance) eventArgs.distance = pointer.distance;
        sceneInfo.dotNetScene.invokeMethodAsync(handlerName, eventArgs);
    });
}

function addSceneEventHandler(sceneName, eventName, handlerName) {
    const sceneInfo = sceneInfos[sceneName];
    sceneInfo.phaserScene.input.on(eventName, (pointer) => {
        const eventArgs = { 'x': pointer.x, 'y': pointer.y, 'distance': -1 };
        //if (pointer.distance) eventArgs.distance = pointer.distance;
        sceneInfo.dotNetScene.invokeMethod(handlerName, eventArgs);
    });
}

function setSpriteCrop(sceneName, spriteName, x, y, width, height) {
    const sprite = findSprite(sceneName, spriteName);
    sprite.setCrop(x, y, width, height);
}

function setSpriteScale(sceneName, spriteName, scale) {
    const sprite = findSprite(sceneName, spriteName);
    sprite.scale = scale;
}

function setSpriteLocation(sceneName, spriteName, x, y) {
    const sprite = findSprite(sceneName, spriteName);
    sprite.x = x;
    sprite.y = y;
}

function spriteExists(sceneName, spriteName) {
    return findSprite(sceneName, spriteName) != null;
}

function setTextValue(sceneName, textName, value) {
    const text = findText(sceneName, textName);
    text.setText(value);
}

function getSpriteData(sceneName, spriteName, key) {
    const sprite = findSprite(sceneName, spriteName);
    return sprite.getData(key);
}

function setSpriteData(sceneName, spriteName, key, value) {
    const sprite = findSprite(sceneName, spriteName);
    sprite.setData(key, value);
}

function addSprite(sceneName, spriteName, imageName, x, y, scale = null, interactive = false) {
    const sceneInfo = sceneInfos[sceneName];
    const sprite = sceneInfo.phaserScene.add.sprite(x, y, imageName);
    sprite.name = spriteName; // TODO Use same strategy for scenes.

    if (scale) {
        sprite.setScale(scale);
    }
}

function addText(sceneName, textName, x, y, text, fontSize, fill) {
    const sceneInfo = sceneInfos[sceneName];
    const textBox = sceneInfo.phaserScene.add.text(x, y, text, { fontSize: fontSize, fill: fill });
    textBox.name = textName;
}

function removeSprite(sceneName, spriteName) {
    const sprite = findSprite(sceneName, spriteName);
    if (sprite) {
        sprite.destroy();
    }
}

function addRectangle(sceneName, x, y, width, height, color) {

    const sceneInfo = sceneInfos[sceneName];
    var graphics = sceneInfo.phaserScene.add.graphics({ fillStyle: { color: color } });

    var rect = new Phaser.Geom.Rectangle(x, y, width, height);

    graphics.fillRectShape(rect);
}

// TODO Rename to active
function isSceneVisible(scene) {
    return game.scene.isActive(scene)
        || sceneInfos[scene].isCreating;
}

function switchScene(from, to) {
    stopScene(from);
    startScene(to);
}

function startScene(scene) {
    game.scene.start(scene);
}

function stopScene(scene) {
    game.scene.stop(scene);
}

function shakeCamera(scene) {
    sceneInfos[scene].phaserScene.cameras.main.flash(250);
}

function registerScene(name, dotNetScene) {

    var phaserScene = new Phaser.Scene(name);
    sceneInfos[name] = {
        phaserScene: phaserScene,
        dotNetScene: dotNetScene
    };

    // TODO Use pack instead of hard coded: https://phaser.io/examples/v3/view/scenes/swapping-scenes
    phaserScene.preload = function () {
        this.load.image('brush', './assets/brush.png');
        this.load.image('logo', './assets/logo.png');
        this.load.image('world', './assets/world.png');
        this.load.image('wally', './assets/wally.png');
        this.load.image('amigo', './assets/amigo.png');
        this.load.image('loading', './assets/loading.gif');
        this.load.image('moodlevel', './assets/moodlevel.png');
    };

    phaserScene.create = function () {

        //this.cameras.add(0, 0, 800, 600);
        sceneInfos[name].isCreating = true;
        dotNetScene.invokeMethod('create');
        sceneInfos[name].isCreating = false;
    }
}

let game;

function startPhaser(container, title) {

    const config = {
        type: Phaser.AUTO,
        parent: container,
        width: 800,
        height: 600,
        scale: {
            mode: Phaser.Scale.WIDTH_CONTROLS_HEIGHT
        },
        scene: Object.keys(sceneInfos).map(key => sceneInfos[key].phaserScene),
        title: title
      };
      
      game = new Phaser.Game(config);
}

